source/pki_utils.c

```diff
@@ -1,298 +1,816 @@
+       https://docs.microsoft.com/en-us/windows/desktop/SecCrypto/system-store-locations
+       this value should be more than enough. We'll add an assert just in case. */
+    char store_path[128] = {0};
+    assert(location_of_next_segment - store_path_start < sizeof(store_path));
+    memcpy(store_path, store_path_start, location_of_next_segment - store_path_start);
+
+    location_of_next_segment += 1;
+    if (strlen(location_of_next_segment) != CERT_HASH_STR_LEN) {
+        return aws_raise_error(AWS_IO_FILE_INVALID_PATH);
+    }
+
+    *cert_store = CertOpenStore(
+        CERT_STORE_PROV_SYSTEM_A, 0, (HCRYPTPROV)NULL, CERT_STORE_OPEN_EXISTING_FLAG | store_val, store_path);
+
+    if (!cert_store) {
+        return aws_raise_error(AWS_IO_FILE_INVALID_PATH);
+    }
+
+    BYTE cert_hash_data[CERT_HASH_LEN];
+    CRYPT_HASH_BLOB cert_hash = {
+        .pbData = cert_hash_data,
+        .cbData = CERT_HASH_LEN,
+    };
+
+    if (!CryptStringToBinaryA(
+            location_of_next_segment,
+            CERT_HASH_STR_LEN,
+            CRYPT_STRING_HEX,
+            cert_hash.pbData,
+            &cert_hash.cbData,
+            NULL,
+            NULL)) {
+        return aws_raise_error(AWS_IO_FILE_INVALID_PATH);
+    }
+
+    *certs = CertFindCertificateInStore(
+        *cert_store, X509_ASN_ENCODING | PKCS_7_ASN_ENCODING, 0, CERT_FIND_HASH, &cert_hash, NULL);
+
+    if (!certs) {
+        aws_close_cert_store(*cert_store);
+        *cert_store = NULL;
+        return aws_raise_error(AWS_IO_FILE_INVALID_PATH);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_import_trusted_certificates(
+    struct aws_allocator *alloc,
+    struct aws_byte_buf *certificates_blob,
+    HCERTSTORE *cert_store) {
+    struct aws_array_list certificates;
+    *cert_store = NULL;
+
+    if (aws_array_list_init_dynamic(&certificates, alloc, 2, sizeof(struct aws_byte_buf))) {
+        return AWS_OP_ERR;
+    }
+
+    if (aws_decode_pem_to_buffer_list(alloc, certificates_blob, &certificates)) {
+        aws_array_list_clean_up(&certificates);
+        return AWS_OP_ERR;
+    }
+
+    int error_code = AWS_OP_SUCCESS;
+    size_t cert_count = aws_array_list_length(&certificates);
+
+    HCERTSTORE tmp_cert_store =
+        CertOpenStore(CERT_STORE_PROV_MEMORY, 0, (ULONG_PTR)NULL, CERT_STORE_CREATE_NEW_FLAG, NULL);
+    *cert_store = tmp_cert_store;
+    if (!*cert_store) {
+        error_code = aws_raise_error(AWS_IO_SYS_CALL_FAILURE);
+        goto clean_up;
+    }
+
+    for (size_t i = 0; i < cert_count; ++i) {
+        struct aws_byte_buf *byte_buf_ptr = NULL;
+        aws_array_list_get_at_ptr(&certificates, (void **)&byte_buf_ptr, i);
+
+        CERT_BLOB cert_blob;
+        CERT_CONTEXT *cert_context = NULL;
+
+        cert_blob.pbData = byte_buf_ptr->buffer;
+        cert_blob.cbData = (DWORD)byte_buf_ptr->len;
+
+        DWORD content_type = 0;
+        BOOL query_res = CryptQueryObject(

```

Comment:
I'm sure it's in the docs, but is this just validating that this is a valid cert blob? Comment perhaps?

Reply:
unfortunately, it also fills in the cert context object for use in the next API call.